### PR TITLE
Electron support

### PR DESCRIPTION
**Pull request by wdanilo**
Implements https://github.com/luna/ide/issues/118

### CODE REVIEW

File: app/build/icon-gen.js [hunk 1]
```diff
@@ -0,0 +1,367 @@
+
```
Comment (farmaazon) [inline app/build/icon-gen.js]: A simple doc explaining what is this file for.

File: app/src/main/index.js [hunk 2]
```diff
@@ -0,0 +1,281 @@
+'use strict'
```
Comment (farmaazon) [inline app/src/main/index.js]: We have `src/index.js` and `main/index.js` for me it's not clear what are the exact responsibilities of the two index.jses 

A little comment at file begin could help a lot.

File: app/src/server.js [hunk 3]
```diff
@@ -0,0 +1,80 @@
+import * as createServers from 'create-servers'
```
Comment (farmaazon) [inline app/src/server.js]: Also add a little doc explaining this file. What is this server? Why we are spawning it?

File: README.md [hunk 4]
```diff
@@ -1,86 +1,52 @@
...
-`scripts/watch.sh --dev` in a second shell. In order to build the IDE application, 
-follow the steps below:
+### Contributing
 
-```bash
-cd app
-npm install
-npm run start 
-```
+Enso language is a community-driven open source project which is and will always be open and free to use. We are committed to a fully transparent development process and highly appreciate every contribution. If you love the vision behind Enso and you want to redefine the data processing world, join us and help us track down bugs, implement new features, improve the documentation or spread the word! Join our community on a Discord chat](http://chat.luna-lang.org) and read the [development and contributing guidelines](https://github.com/luna/ide/docs/contributing.md).
```
Comment (BinarySoftware) [inline README.md]: A small typo : "  Discord chat](http://chat.luna-lang.org) " -> " [Discord chat](http://chat.luna-lang.org)

File: README.md [hunk 5]
```diff
@@ -1,86 +1,77 @@
...
+</p>
+
+Enso is a visual data science language selected by Singularity University and NASA as one of 20 most 
+unique technologies worldwide.
+
+<br/>
+
+### Getting Started
+
+[Download a pre-build package for MacOS, Windows, or Linux](https://github.com/luna/ide/releases),
```
[the remark this answers is not in the document]
Comment (wdanilo) [reply README.md]: It is part of this task. The docs are changed to show the shape - we do not have all the links now.

File: README.md [hunk 6]
```diff
@@ -1,86 +1,77 @@
...
+contribution. If you love the vision behind Enso and you want to redefine the data processing world, 
+join us and help us track down bugs, implement new features, improve the documentation or spread the 
+word! Join our community on a Discord chat](http://chat.luna-lang.org) and read the 
+[development and contributing guidelines](https://github.com/luna/ide/docs/contributing.md).
+
+<br/>
+
+### License
+
+The Enso Language Compiler is released under the terms of the [Apache v2 License](). The Enso
```
[the remark this answers is not in the document]
Comment (wdanilo) [reply README.md]: See the reply above

File: README.md [hunk 7]
```diff
@@ -1,86 +1,77 @@
...
+join us and help us track down bugs, implement new features, improve the documentation or spread the 
+word! Join our community on a Discord chat](http://chat.luna-lang.org) and read the 
+[development and contributing guidelines](https://github.com/luna/ide/docs/contributing.md).
+
+<br/>
+
+### License
+
+The Enso Language Compiler is released under the terms of the [Apache v2 License](). The Enso 
+Graphical Interface and it's rendering engine are released under the terms of the [AGPL License]().
```
[the remark this answers is not in the document]
Comment (wdanilo) [reply README.md]: See the reply above

File: build/run.js [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+async function clean_js () {
+    await cmd.with_cwd('app', async () => {
+        await cmd.run('npm',['run','clean'].concat(child_argv))
+    })
+    try { await fs.unlink('.initialized') } catch {}
+}
+
+async function clean_rust () {
+    try { await fs.rmdir('app/generated') } catch {}
+    // TODO finish
```
Comment (farmaazon) [inline build/run.js]: I think we should do also `cargo clean` to remove target directory.

File: docs/contributing.md [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+  usage.
+  
+- **Testing**
+  Use `run test` run both unit and web-based test.
+  
+<br/>
+
+### Testing
+The sources use both unit tests and web test, which are run in a browser and
+produce visual results. To run them, use the `scripts/test.sh` script and follow
```
Comment (mwu-tow) [inline docs/contributing.md]: Here and further — scripts -> build
Also, it is really unclear how different is this from `run test`.

File: docs/contributing.md [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+  rustup toolchain install nightly-2019-11-04 # Install the nightly channel.
+  rustup default nightly                      # Set it as the default one.
+  rustup component add clippy                 # Install the linter.
+  ```
+
+- **Node Version Manager**  
+  In order to build the web and desktop applications you will need [node](https://nodejs.org) and 
+[npm](https://www.npmjs.com). Please note that the versions available in your system package manager
+will probably crash while building native-extensions. The only known stable solution is to use the 
+[Node Version Manager](https://github.com/nvm-sh/nvm). Please note that installing it from any
```
Comment (mwu-tow) [inline docs/contributing.md]: TThe NVM explicitly states that it doesn't support Windows. Alternatives should be explored. Or consequences of not using it properly explained.

File: docs/contributing.md [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+Please be sure to fix all errors reported by `scripts/lint.sh` before creating a
+pull request to this repository.
+
+<br/>
+<br/>
+
+## Distribution
+
+### Building the Native Application
+Enter the `app` directory, run `nvm use` to set up the correct node environment and run `npm run electron:dev` to build and start a native application in a development mode.
```
Comment (mwu-tow) [inline docs/contributing.md]: I don't have `nvm`,

File: run [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+            cmd.section('Validation')
+            if (!args.includes(no_nvm_check)) {
+                if (process.env.NVM_DIR == undefined) {
+                    throw nvm_error
+                }
+                // TODO[WD]: Why its not working?
+                // await cmd.check_version('nvm','0.35.2')
+            }
+            await cmd.check_version('npm','6.14.1')
+            await cmd.check_version('node','v12.16.1')
```
Comment (mwu-tow) [inline run]: I don't think that current CI script ensures that we get such version. If so, any external update to actions/images will break us. 
Either CI script should be strengthened or this checks weakened.

File: docs/contributing.md [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+below:
+
+- **Interactive builds**  
+  Run `./run watch` to start a local web-server and a source-file watch utility which will build the 
+  project on every change. Open `http://localhost:8080` to run the application, or 
+  `http://localhost:8080/debug` to see example demo scenes. Please remember to disable the cache in 
+  your browser during development! By default, the script disables heavyweight optimizations to 
+  provide interactive development experience. The scripts are thin wrappers for 
+  [wasm-pack](https://github.com/rustwasm/wasm-pack) and accept the same 
+  [command line arguments](https://rustwasm.github.io/wasm-pack/book/commands/build.html).
```
Comment (mwu-tow) [inline docs/contributing.md]: I don't think it is true anymore?
Or please show example of passing arguments through the new wrappers.

File: docs/contributing.md [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+- **Node Version Manager**  
+  In order to build the web and desktop applications you will need [node](https://nodejs.org) and 
+[npm](https://www.npmjs.com). Please note that the versions available in your system package manager
+will probably crash while building native-extensions. The only known stable solution is to use the 
+[Node Version Manager](https://github.com/nvm-sh/nvm). Please note that installing it from any 
+package manager is officially not supported and can cause issues. Follow the 
+[official guide](https://github.com/nvm-sh/nvm#installing-and-updating) to install it:
+
+  ```bash
+  curl -o- https://raw.githubusercontent.com/nvm-sh/nvm/v0.35.2/install.sh | bash
```
[the remark this answers is not in the document]
Comment (wdanilo) [reply docs/contributing.md]: This is copy-paste from `nvm` docs. Refer to them to learn how to do it on widnows then :(

File: docs/contributing.md [hunk 15]
```diff
@@ -0,0 +1,121 @@
...
+
+<br/>
+<br/>
+<br/>
+
+## Working with sources
+As this is a multi-part project with many complex dependencies, it was equipped with a build script
+which both validates your working environment as well as takes care of providing most suitable 
+compilation flags for a particular development stage. In order to learn more about the commands and 
+available options, simply run `./run` and read the manual. The most common options are presented
```
[the remark this answers is not in the document]
Comment (wdanilo) [reply docs/contributing.md]: I'll add note about windows

File: docs/contributing.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (wdanilo) [reply docs/contributing.md]: Ok, I explained the situation better in the docs

**Comment by wdanilo** [comment]
CI fails because of this : https://github.com/withoutboats/failure_derive/issues/13
As agreed with MWU, merging.

**Pull request merged**